src/cmd/troubleshoot/dynakube.go

```diff
@@ -0,0 +1,280 @@
+	logInfof("checking if api url is valid")
+
+	query := kubeobjects.NewDynakubeQuery(nil, troubleshootCtx.apiReader, troubleshootCtx.namespaceName).WithContext(context.TODO())
+	dynakube, err := query.Get(types.NamespacedName{Namespace: troubleshootCtx.namespaceName, Name: troubleshootCtx.dynakubeName})
+	if err != nil {
+		logWithErrorf(err, "selected '%s:%s' Dynakube does not exist", troubleshootCtx.namespaceName, troubleshootCtx.dynakubeName)
+		return err
+	}
+
+	validation.SetLogger(log)
+	if validation.NoApiUrl(nil, &dynakube) != "" {
+		logErrorf("api url is invalid")
+		return fmt.Errorf("api url is invalid")
+	}
+	if validation.IsInvalidApiUrl(nil, &dynakube) != "" {
+		logErrorf("api url is invalid")
+		return fmt.Errorf("api url is invalid")
+	}
+
+	logInfof("api url is valid")
+	return nil
+}
+
+func evaluateDynatraceApiSecretName(troubleshootCtx *troubleshootContext) error {
+	logInfof("checking if secret is valid")
+
+	// use dynakube name or tokens value if set
+	troubleshootCtx.dynatraceApiSecretName = troubleshootCtx.dynakubeName
+
+	query := kubeobjects.NewDynakubeQuery(nil, troubleshootCtx.apiReader, troubleshootCtx.namespaceName).WithContext(context.TODO())
+	dynakube, err := query.Get(types.NamespacedName{Namespace: troubleshootCtx.namespaceName, Name: troubleshootCtx.dynakubeName})
+	if err != nil {
+		logWithErrorf(err, "selected '%s:%s' Dynakube does not exist", troubleshootCtx.namespaceName, troubleshootCtx.dynakubeName)
+		return err
+	}
+
+	if dynakube.Spec.Tokens != "" {
+		troubleshootCtx.dynatraceApiSecretName = dynakube.Spec.Tokens
+	}
+	return nil
+}
+
+func checkIfDynatraceApiSecretWithTheGivenNameExists(troubleshootCtx *troubleshootContext) error {
+	query := kubeobjects.NewSecretQuery(context.TODO(), nil, troubleshootCtx.apiReader, log)
+	if _, err := query.Get(types.NamespacedName{Namespace: troubleshootCtx.namespaceName, Name: troubleshootCtx.dynatraceApiSecretName}); err != nil {
+		logWithErrorf(err, "'%s:%s' secret is missing", troubleshootCtx.namespaceName, troubleshootCtx.dynatraceApiSecretName)
+		return err
+	}
+	logInfof("'%s:%s' secret exists", troubleshootCtx.namespaceName, troubleshootCtx.dynatraceApiSecretName)
+	return nil
+}
+
+func checkIfDynatraceApiSecretHasApiToken(troubleshootCtx *troubleshootContext) error {
+	query := kubeobjects.NewSecretQuery(context.TODO(), nil, troubleshootCtx.apiReader, log)
+	secret, err := query.Get(types.NamespacedName{Namespace: troubleshootCtx.namespaceName, Name: troubleshootCtx.dynatraceApiSecretName})
+	if err != nil {
+		logWithErrorf(err, "'%s:%s' secret is missing", troubleshootCtx.namespaceName, troubleshootCtx.dynatraceApiSecretName)
+		return err
+	}
+
+	apiTokenByte, ok := secret.Data["apiToken"]
+	if !ok {
+		logErrorf("token apiToken does not exist in '%s:%s' secret", troubleshootCtx.namespaceName, troubleshootCtx.dynatraceApiSecretName)
+		return fmt.Errorf("token apiToken does not exist in secret")
```

Comment:
why do we log it and return the error in a different format? should keep consistency